util/util.go

```diff
@@ -51,3 +58,62 @@ func MakeSignature(size int) ([]byte, error) {
 	b = []byte(base64.StdEncoding.EncodeToString(b))
 	return b, err
 }
+
+// RetryUntilSucceed keep retrying given action with specified timeout until action succeed or specified context is done.
+func RetryUntilSucceed(action func() error, desc string, ctx context.Context, timeout time.Duration) {
+	ctxCompleted := false
+	go func() {
+		select {
+		case <-ctx.Done():
+			ctxCompleted = true
+		}
+	}()
+	for {
+		err := action()
+		if err == nil {
+			return
+		}
+		if err != nil {
+			if ctxCompleted {
+				log.Infof("Stop retrying %s", desc)
+				return
+			}
+			log.Warnf("Failed to %s: %+v, retrying in %v", desc, err, timeout)
+			time.Sleep(timeout)
+		}
+
+	}
+}
+
+func FirstNonEmpty(args ...string) string {
+	for _, value := range args {
+		if len(value) > 0 {
+			return value
+		}
+	}
+	return ""
+}
+
+func RunAllAsync(count int, action func(i int) error) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			message := fmt.Sprintf("Recovered from panic: %+v\n%s", r, debug.Stack())
+			log.Error(message)
+			err = errors.New(message)
+		}
+	}()
+	var wg sync.WaitGroup
+	for i := 0; i < count; i++ {
+		index := i
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			err = action(index)
+		}()
```

Comment:
Theres a race condition here. The value of index will change during iteration. the value of `i` should be passed as an argument to the goroutine.

Reply:
Thank you! fixed: https://github.com/argoproj/argo-cd/pull/840/commits/a842eb3bb31f378c37866ea5deba3efc150d4341